app/helpers/merge_requests_helper.rb

```diff
@@ -104,10 +104,74 @@ def link_to_merging_organisation_users(organisation)
   end
 
   def total_users_after_merge_text(merge_request)
-    count = merge_request.total_visible_users_after_merge
+    count = merge_request.total_visible_users_after_merge || 0
     "#{"#{count} user".pluralize(count)} after merge"
   end
 
+  def total_stock_owners_after_merge_text(merge_request)
+    count = merge_request.total_stock_owners_after_merge || 0
+    "#{"#{count} stock owner".pluralize(count)} after merge"
+  end
+
+  def total_managing_agents_after_merge_text(merge_request)
+    count = merge_request.total_managing_agents_after_merge || 0
+    "#{"#{count} managing agent".pluralize(count)} after merge"
+  end
+
+  def stock_owners_text(merge_request)
+    relationship_text(merge_request, "stock_owner", :stock_owners_organisation_path)
+  end
+
+  def managing_agent_text(merge_request)
+    relationship_text(merge_request, "managing_agent", :managing_agents_organisation_path)
+  end
+
+  def relationship_text(merge_request, relationship_type, relationship_path_helper)
+    html_content = ""
+    duplicates = duplicate_relationships(merge_request, relationship_type)
+    organisations_without_relationships = ([merge_request.absorbing_organisation] + merge_request.merging_organisations).select { |org| org.send(relationship_type.pluralize).visible.empty? }
+
+    if duplicates.any?
+      html_content += "Some of the organisations merging have common #{relationship_type.humanize(capitalize: false).pluralize}.<br><br>"
+    end
+
+    if organisations_without_relationships.any?
+      org_names = organisations_without_relationships.map(&:name).to_sentence
+      verb = organisations_without_relationships.count > 1 ? "have" : "has"
+      html_content += "#{org_names} #{verb} no #{relationship_type.humanize(capitalize: false).pluralize}.<br><br>"
+    end
+
+    ([merge_request.absorbing_organisation] + merge_request.merging_organisations).each do |org|
+      relationship_count = org.send(relationship_type.pluralize).visible.count
+      next if relationship_count.zero?
+
+      link_text = if relationship_count == 1
+                    "View the #{relationship_count} #{org.name} #{relationship_type.humanize(capitalize: false)} (opens in a new tab)"
+                  else
+                    "View all #{relationship_count} #{org.name} #{relationship_type.humanize(capitalize: false).pluralize} (opens in a new tab)"
+                  end
+      html_content += "#{govuk_link_to(link_text, send(relationship_path_helper, org), target: '_blank')}<br><br>"
+    end
+
+    html_content.html_safe
+  end
+
+  def duplicate_relationships(merge_request, relationship_type)
+    relationships = merge_request.absorbing_organisation.send(relationship_type.pluralize) + merge_request.merging_organisations.flat_map { |org| org.send(relationship_type.pluralize) }
```

Comment:
Maybe call this `related_organisations` instead of `relationships`? I'm assuming this returns orgs?

Reply:
Good idea that explains better

Reply:
This method is supposed to be reused for both stock owners and managing agents